Added Python efr algorithm implementation

Jamesflynn1 opened the pull request:
A Python implementation of the EFR (https://arxiv.org/abs/2102.06973) algorithm with the deviation types defined in the proposing paper. The implementation was developed as part of my undergraduate dissertation and I thought it would be beneficial if the implementation could be made available as part of OpenSpiel.

Uses a shortcut for external only deviations as found in https://github.com/dmorrill10/hr_edl_experiments (original experimental code)

Any guidance or feedback would be appreciated as this is my first contribution to an open source project.

Todo:

- [x] Look into issues with behavioural deviations
- [x] Look into possible incorrect definition with CSPS deviations
- [x] Add docstrings for all functions
- [x] Remove discounting code
- [x] Refactor code
- [x] Ensure adherence to Google Python style
- [x] Add updates to OpenSpiel docs

lanctot commented:
Thanks @Jamesflynn1 !

Please let me know when it is ready for a review.

lanctot commented on open_spiel/python/algorithms/efr.py:
```diff
@@ -0,0 +1,967 @@
+# Modified: 2023 James Flynn
+# Original: https://github.com/deepmind/open_spiel/blob/master/open_spiel/python/algorithms/cfr.py
+"""Python implementation of the extensive-form regret minimization algorithm.
+
```
Can you add a reference to the paper here?

lanctot commented on open_spiel/python/algorithms/efr.py:
```diff
@@ -0,0 +1,967 @@
+
+@attr.s
+class _InfoStateNode(object):
+    """An object wrapping values associated to an information state."""
```
Throughout OpenSpiel, we use two spaces for indentation. Can you fix this and the other file to make it consistent?

Also can you run pylint with Google style files? (See step 9 of Adding a New Game here: https://github.com/google-deepmind/open_spiel/blob/master/docs/developer_guide.md#adding-a-game)

Jamesflynn1 replied:
Thanks for the comments!

I ran both files through pylint with the style guide and added a few more comments.

Let me know what you think, I am more than happy to make changes/clarify anything as required.

lanctot commented on open_spiel/python/algorithms/efr_test.py:
```diff
@@ -0,0 +1,69 @@
+from open_spiel.python.algorithms import expected_game_score
+import pyspiel
+
+_KUHN_GAME = pyspiel.load_game("kuhn_poker")
```
Don't make these global. You can put them into the object, by initializing them in a setup function in the EFRTest class. The signature is:

```
def setUp(self):
```

lanctot commented on open_spiel/python/algorithms/efr_test.py:
```diff
@@ -0,0 +1,69 @@
+_LEDUC_UNIFORM_POLICY = policy.TabularPolicy(_LEDUC_GAME)
+_DEVIATIONS_ = ["blind action", "informed action", "blind cf", "informed cf", "bps", "cfps", "csps", "tips", "bhv"]
+
+class EFRTest(parameterized.TestCase, absltest.TestCase):
```
Since this is EFR, can we also have a non-zero sum game and multi-player game?

We've exposed some of the distance to the various *CE and *CCE algorithm. I just noticed that we don't have examples of using them, though, I'll add some and let you know.

lanctot replied:
I added some examples of checking distance to CCE here: https://github.com/google-deepmind/open_spiel/commit/8859b79f895ce7bddaf62d77ba6c0946956312ea. Should be easy to expose the other ones (EF(C)CE, AF(C)CE) if you want to try them too. 

Would be good to have at least one graph where we confirm EFR approaching is CE or CCE when it should be, and then a test on a smallish game where this distance is decreasing (over a small number of iterations so the test doesn't take too long -- see the CFR tests I added).

Jamesflynn1 replied:
Amazing, thanks for the suggestions, I'll work on this over the next few days and let you know if I run into any issues.

Jamesflynn1 replied:
I have ran a test on 3 player Kuhn Poker for all deviation types over 1000 iterations and plotted the distance to the CCE.
![CCE restricted](https://github.com/google-deepmind/open_spiel/assets/13368287/e9fc3abd-baaa-412e-b601-84299b12a4c1)
![CCE full deviations](https://github.com/google-deepmind/open_spiel/assets/13368287/5ac9ce08-a812-4257-b55e-8d954a7a431d)
 I think it roughly checks out, let me know what you think!

lanctot commented on open_spiel/python/algorithms/efr.py:
```diff
@@ -0,0 +1,1006 @@
+
+The average policy converges to a Nash Equilibrium rather than the current policy.
+"""
+import copy
```
add a space here

lanctot commented on open_spiel/python/algorithms/efr.py:
```diff
@@ -0,0 +1,1006 @@
+The average policy converges to a Nash Equilibrium rather than the current policy.
+"""
+import copy
+from collections import defaultdict
```
Please run this through the Google Python style and fix the lint errors. Use [pylint](https://pypi.org/project/pylint/) with the [pylintrc from the Google style guide](https://google.github.io/styleguide/pyguide.html)

lanctot commented on open_spiel/python/algorithms/efr.py:
```diff
@@ -0,0 +1,1006 @@
+  history = attr.ib()
+
+  cumulative_regret = attr.ib(factory=lambda: defaultdict(float))
+  #The sum of all prior iteration's policies
```
Add a space after the "#"

lanctot commented on open_spiel/python/algorithms/efr.py:
```diff
@@ -0,0 +1,1006 @@
+@attr.s
+class _InfoStateNode(object):
+  """An object wrapping values associated to an information state."""
+  # The list of the legal actions.
```
Blank line needed after comment (these will all be caught by the linter with Google python style)

lanctot commented on open_spiel/python/algorithms/efr.py:
```diff
@@ -0,0 +1,1006 @@
+  def __init__(self, game, deviation_gen):
+    """Initializer.
+    Args:
+    game: The `pyspiel.Game` to run on.
```
Lines have to be under 80 characters

lanctot reviewed (changes requested):
Please fix the style to conform to Google Python style. I've pointed out a few examples, but please  fix them all.

lanctot commented:
Hi @Jamesflynn1, if any of your recent commits have addressed the comments, can you reply to them and/or mark them as resolved?

Jamesflynn1 commented:
> Hi @Jamesflynn1, if any of your recent commits have addressed the comments, can you reply to them and/or mark them as resolved?

Hi @lanctot,  apologies I haven't replied sooner, I have been very busy with university coursework and graduate applications. 

I ran pylint with the Google style guide and I was still some of style issues (spaces after comments and blank lines), I'll take another look now and mark off the solved issues.

Jamesflynn1 commented:
Hi @lanctot,

I have tried to refractor the files using the linked pylintrc file and I'm not getting all the commenting suggestions that has been pointed out. I've fixed (hopefully) all instances of the issues you have pointed out. However, I think it's possible that I have a different pylintrc file (I get no linting issues when I pylint the file) so to catch all issues would it be possible to compare files?

Any assistance would be appreciated!

Thanks,
James

lanctot commented:
No problem, I can try to fix these on our side. Apologies for the delays.

lanctot reviewed (changes requested):
Can you add efr_test.py to the python/CMakeLists.txt

We need to ensure that it runs on Github Actions CI before we import it. 

You might need to pull changes from master because that file has changed a few times since our last contact.
